### PR TITLE
refactor(terraform-provider): convert to class-based datasource

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes:

Refactor the `terraform-provider` datasource to be class-based.

## Context:

Related to #8647

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/terraform-provider/index.ts [hunk 1]
```diff
@@ -1,119 +1,116 @@
...
-    return null;
+  @cache({
+    namespace: `datasource-${TerraformProviderDatasource.id}`,
+    key: (getReleasesConfig: GetReleasesConfig) =>
+      `${getReleasesConfig.registryUrl}/${
+        getReleasesConfig.lookupName.includes('/')
+          ? getReleasesConfig.lookupName
+          : `hashicorp/${getReleasesConfig.lookupName}`
+      }`,
+  })
```
Comment (JamieMagee) [inline lib/datasource/terraform-provider/index.ts:37]: I'm on the fence about this. On one hand I know it's a bit difficult to read, on the other I'm not sure it's _that_ bad to be worth extracting to a separate function.

Comment (viceice) [reply lib/datasource/terraform-provider/index.ts:37]: It's used two times, so i think it's s good candidate for extraction

**Review by viceice — COMMENTED** [review]
Can you check if this collides with #8429?

**Comment by JamieMagee** [comment]
> Can you check if this collides with #8429?

It appears not to conflict

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**